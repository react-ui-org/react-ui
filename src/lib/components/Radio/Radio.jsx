import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import styles from './Radio.scss';

export const Radio = ({
  disabled,
  helpText,
  id,
  isLabelVisible,
  label,
  layout,
  options,
  required,
  validationState,
  validationText,
  value,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <div
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.rootLayoutHorizontal
          : styles.rootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      id={id}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <ul className={styles.list}>
          {
            options.map((option) => (
              <li key={option.value}>
                <label
                  className={styles.option}
                  htmlFor={id && `${id}__item__${option.value}`}
                  id={id && `${id}__item__${option.value}__label`}
                >
                  <input
                    {...transferProps(restProps)}
                    className={styles.input}
                    checked={restProps.onChange
                      ? (value === option.value) || false
                      : undefined}
                    disabled={disabled || option.disabled}
                    id={id && `${id}__item__${option.value}`}
                    name={id}
                    type="radio"
                    value={option.value}
                  />
                  <span
                    className={styles.optionLabel}
                    id={id && `${id}__item__${option.value}__labelText`}
                  >
                    { option.label }
                  </span>
                </label>
              </li>
            ))
          }
        </ul>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </div>
  );
};

Radio.defaultProps = {
  disabled: false,
  helpText: null,
  id: undefined,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  validationState: null,
  validationText: null,
  value: undefined,
};

Radio.propTypes = {
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the root HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   *
   * and of individual options (`<input>`):
   * * `<ID>__item__<VALUE>`
   * * `<ID>__item__<VALUE>__label`
   * * `<ID>__item__<VALUE>__labelText`
   */
  id: PropTypes.string,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Label of the group of options.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * Set of options to be chosen from.
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    disabled: PropTypes.bool,
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export const RadioWithContext = withProviderContext(Radio, 'Radio');

export default RadioWithContext;
