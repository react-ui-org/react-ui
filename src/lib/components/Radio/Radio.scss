@use "../../styles/tools/form-fields/box-field-elements";
@use "../../styles/tools/form-fields/box-field-layout";
@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/inline-field-elements";
@use "../../styles/tools/form-fields/inline-field-layout";
@use "../../styles/tools/form-fields/variants";
@use "../../styles/tools/accessibility";
@use "../../styles/tools/reset";
@use "../../styles/tools/spacing";

// Foundation
.root {
    @include foundation.root();
    @include variants.visual(check);
}

.label,
.optionLabel {
    @include foundation.label();
}

.list {
    @include reset.list();
}

.option {
    @include inline-field-layout.field();
    @include inline-field-elements.min-tap-target($type: radio);
}

.input {
    @include inline-field-elements.check-input($type: radio);
}

.helpText,
.validationText {
    @include foundation.help-text();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
    @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
    @include box-field-layout.vertical($has-list: true);
}

.rootLayoutHorizontal {
    @include box-field-layout.horizontal($has-min-tap-target: true);
}

.isRootFullWidth {
    @include box-field-layout.full-width();
}

.isRootInFormLayout {
    @include box-field-layout.in-form-layout();
}
