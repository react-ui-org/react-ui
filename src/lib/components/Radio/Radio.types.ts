export interface RadioProps {
  /**
   * Rest of the props.
   */
  [key: string]: unknown;
  /**
   * If `true`, the input will be disabled.
   */
  disabled?: boolean;
  /**
   * Optional help text.
   */
  helpText?: React.ReactNode;
  /**
   * ID of the root HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   *
   * and of individual options (`<input>`):
   * * `<ID>__item__<VALUE>`
   * * `<ID>__item__<VALUE>__label`
   * * `<ID>__item__<VALUE>__labelText`
   *
   * If `key` in the option definition object is set,
   * then `option.key` is used instead of `option.value` in place of `<VALUE>`.
   */
  id?: string;
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible?: boolean;
  /**
   * Label of the group of options.
   */
  label: string;
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   */
  layout?: Layout;
  /**
   * Set of options to be chosen from.
   *
   * For generating unique IDs the `option.value` is normally used. For cases when this is not practical or
   * the `option.value` values are not unique the `option.key` attribute can be set manually.
   */
  options: {
    disabled?: boolean;
    key?: string;
    label: string;
    value?: string | number;
  }[];
  /**
   * If `true`, the input will be required.
   */
  required?: boolean;
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState?: Validation;
  /**
   * Validation message to be displayed.
   */
  validationText?: React.ReactNode;
  /**
   * Value of the input.
   */
  value?: string | number;
}
