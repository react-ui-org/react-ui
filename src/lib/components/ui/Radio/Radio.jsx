import PropTypes from 'prop-types';
import React from 'react';
import transferProps from '../../../utils/transferProps';
import styles from './Radio.scss';

export const Radio = (props) => {
  let labelClass = styles.label;
  let rootInFormLayoutClass = '';
  let rootLayoutClass = '';
  let rootValidationStateClass = '';

  if (props.isLabelVisible) {
    if (props.required) {
      labelClass = styles.isLabelRequired;
    }
  } else {
    labelClass = styles.isLabelHidden;
  }

  if (props.inFormLayout) {
    rootInFormLayoutClass = styles.isRootInFormLayout;
  }

  if (props.layout === 'horizontal') {
    rootLayoutClass = styles.isRootLayoutHorizontal;
  }

  if (props.validationState === 'invalid') {
    rootValidationStateClass = styles.isRootStateInvalid;
  } else if (props.validationState === 'valid') {
    rootValidationStateClass = styles.isRootStateValid;
  } else if (props.validationState === 'warning') {
    rootValidationStateClass = styles.isRootStateWarning;
  }

  const propsToTransfer = transferProps(
    props,
    ['changeHandler', 'description', 'disabled', 'error', 'id', 'inFormLayout', 'isLabelVisible',
      'label', 'layout', 'options', 'required', 'validationState', 'value'],
  );

  return (
    <div
      className={`
        ${styles.root}
        ${rootInFormLayoutClass}
        ${rootLayoutClass}
        ${rootValidationStateClass}
      `.trim()}
    >
      <div
        className={labelClass}
        id={`${props.id}__labelText`}
      >
        {props.label}
      </div>
      <ul className={styles.list}>
        {
          props.options.map((option) => (
            <li key={option.value}>
              { /* Rule is deprecated: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/label-has-for.md */ }
              { /* eslint-disable-next-line jsx-a11y/label-has-associated-control */ }
              <label
                className={styles.inputWrap}
                id={`${props.id}__item__${option.value}__label`}
              >
                <input
                  {...propsToTransfer}
                  id={`${props.id}__item__${option.value}`}
                  name={props.id}
                  type="radio"
                  value={option.value}
                  onChange={props.changeHandler}
                  className={styles.input}
                  disabled={props.disabled || option.disabled}
                  checked={(props.value === option.value) || false}
                />
                <span
                  className={styles.radioLabel}
                  id={`${props.id}__item__${option.value}__labelText`}
                >
                  { option.label }
                </span>
              </label>
            </li>
          ))
        }
      </ul>
      {props.description && (
        <div
          className={styles.description}
          id={`${props.id}__descriptionText`}
        >
          {props.description}
        </div>
      )}
      {props.error && (
        <div
          className={styles.error}
          id={`${props.id}__errorText`}
        >
          {props.error}
        </div>
      )}
    </div>
  );
};

Radio.defaultProps = {
  changeHandler: null,
  description: null,
  disabled: false,
  error: null,
  inFormLayout: false,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  validationState: null,
  value: undefined,
};

Radio.propTypes = {
  /**
   * Function to call when the input has changed.
   */
  changeHandler: PropTypes.func,
  /**
   * Optional description.
   */
  description: PropTypes.string,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Error message to be displayed.
   */
  error: PropTypes.string,
  /**
   * Prefix for ID of important inner elements: `<ID>__labelText`, `<ID>__descriptionText`,
   * `<ID>__errorText`, and all options: `<ID>__item__<VALUE>` (individual inputs),
   * `<ID>__item__<VALUE>__label`, and `<ID>__item__<VALUE>__labelText`.
   */
  id: PropTypes.string.isRequired,
  /**
   * Treat the field differently when it's inside a FormLayout. Do not set manually!
   */
  inFormLayout: PropTypes.bool,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Label of the group of options.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field. It has impact only on Radio inside a FormLayout.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * Set of options to be chosen from.
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    disabled: PropTypes.bool,
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default Radio;
