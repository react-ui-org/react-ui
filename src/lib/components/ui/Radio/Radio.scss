@use '../../../styles/tools/forms/box-field-elements';
@use '../../../styles/tools/forms/box-field-layout';
@use '../../../styles/tools/forms/foundation';
@use '../../../styles/tools/forms/inline-field-layout';
@use '../../../styles/tools/forms/relationships';
@use '../../../styles/tools/forms/variants';
@use '../../../styles/tools/accessibility';
@use '../../../styles/tools/reset';
@use '../../../styles/tools/spacing';

// Foundation
.root {
  @include foundation.root();
  @include variants.visual(neutral);
  @include relationships.horizontal-neighbor();
}

.label,
.optionLabel {
  @include foundation.label();
}

.list {
  @include reset.list();
}

.option {
  @include inline-field-layout.field();
}

.helpText,
.validationText {
  @include foundation.help-text();
}

.isRootRequired .label {
  @include foundation.label-required();
}

// States
.isRootStateInvalid {
  @include variants.feedback(invalid);
}

.isRootStateValid {
  @include variants.feedback(valid);
}

.isRootStateWarning {
  @include variants.feedback(warning);
}

// Invisible label
.isLabelHidden {
  @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
  @include box-field-layout.vertical($has-list: true);
}

.rootLayoutHorizontal {
  @include box-field-layout.horizontal();
}

.isRootFullWidth {
  @include box-field-layout.full-width();
  @include relationships.vertical-neighbor();
}

.isRootInFormLayout {
  @include box-field-layout.in-form-layout();
}
