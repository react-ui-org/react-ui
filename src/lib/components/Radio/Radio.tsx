import React, { useContext } from 'react';
import { withGlobalProps } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import styles from './Radio.scss';
import { RadioProps } from './Radio.types';

export const Radio: React.FunctionComponent<RadioProps> = ({
  disabled = false,
  helpText,
  id,
  isLabelVisible = true,
  label,
  layout = 'vertical',
  options,
  required = false,
  validationState,
  validationText,
  value,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <div
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.isRootLayoutHorizontal
          : styles.isRootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      id={id}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <ul className={styles.list}>
          {
            options.map((option) => {
              const key = option.key ?? option.value;
              return (
                <li key={key}>
                  <label
                    className={styles.option}
                    htmlFor={id && `${id}__item__${key}`}
                    id={id && `${id}__item__${key}__label`}
                  >
                    <input
                      {...transferProps(restProps)}
                      checked={Object.prototype.hasOwnProperty.call(restProps, 'onChange')
                        ? (value === option.value) || false
                        : undefined}
                      className={styles.input}
                      disabled={disabled || option.disabled}
                      id={id && `${id}__item__${key}`}
                      name={id}
                      type="radio"
                      value={option.value}
                    />
                    <span
                      className={styles.optionLabel}
                      id={id && `${id}__item__${key}__labelText`}
                    >
                      {option.label}
                    </span>
                  </label>
                </li>
              );
            })
          }
        </ul>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </div>
  );
};

export const RadioWithGlobalProps = withGlobalProps(Radio, 'Radio');

export default RadioWithGlobalProps;
