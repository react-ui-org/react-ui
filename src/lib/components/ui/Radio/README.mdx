---
name: Radio
menu: 'UI'
route: /components/ui/radio
---

# Radio

Radio allows users to select one option from a set.

import {
  Playground,
  Props,
} from 'docz'
import { Radio } from './Radio'

## Basic Usage

To implement the Radio component, you need to import it first:

```js
import { Radio } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    return (
      <Radio
        changeHandler={(e) => setFruit(e.target.value)}
        id="my-radio"
        label="Your favourite fruit"
        options={[
          {
            label: 'Apple',
            value: 'apple',
          },
          {
            label: 'Banana',
            value: 'banana',
          },
          {
            label: 'Grapefruit',
            value: 'grapefruit',
          },
        ]}
        value={fruit}
      />
    );
  }}
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use Radio for **just a few options**. For larger sets of many options (say 4
  and more) consider using the [SelectField](/components/ui/select-field)
  component. This will help keep your UI clean and uncluttered and prevent your
  users from being overwhelmed by too many options.

- **Don't use for boolean** (true/false) selection or to toggle things on and
  off. [CheckboxField](/components/ui/checkbox-field) and
  [Toggle](/components/ui/toggle) are more suitable for such cases.

- Use **short and descriptive labels**, ideally nouns rather than seemingly
  polite phrases like _Please select your favourite fruit_. Short labels will
  help your users accomplish their task faster.

- Only make the Radio's label invisible when there is **another visual
  clue** to guide users through filling the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

ðŸ“– [Read more about checkboxes and radios at Nielsen Norman Group.](https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/)

## Invisible Label

While it may be acceptable for login screens with just a few fields or for other
simple forms, it's dangerous to hide labels from users in most cases. Keep in
mind you should **provide another visual clue** so users know what to fill into
the input.

<Playground>
  {() => {
    const [frequency, setFrequency] = React.useState('weekly');
    return (
      <Radio
        changeHandler={(e) => setFrequency(e.target.value)}
        id="my-radio-hidden-label"
        isLabelVisible={false}
        label="Newsletter frequency"
        options={[
          {
            label: 'I want to subscribe to the weekly newsletter',
            value: 'weekly',
          },
          {
            label: 'I want to subscribe to the monthly newsletter',
            value: 'monthly',
          },
          {
            label: "I don't wish to receive anything",
            value: 'never',
          },
        ]}
        value={frequency}
      />
    );
  }}
</Playground>

## Horizontal Layout

The default vertical layout is very easy to use and work with. However, there
are situations where horizontal layout suits better â€” and that's why React UI
supports this kind of layout as well.

<Playground>
  {() => {
    const [frequency, setFrequency] = React.useState('weekly');
    return (
      <Radio
        changeHandler={(e) => setFrequency(e.target.value)}
        id="my-radio-horizontal"
        label="Newsletter frequency"
        layout="horizontal"
        options={[
          {
            label: 'I want to subscribe to the weekly newsletter',
            value: 'weekly',
          },
          {
            label: 'I want to subscribe to the monthly newsletter',
            value: 'monthly',
          },
          {
            label: "I don't wish to receive anything",
            value: 'never',
          },
        ]}
        value={frequency}
      />
    );
  }}
</Playground>

## Help Text

You may provide an additional help text to clarify how the input should be
filled.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    return (
      <Radio
        changeHandler={(e) => setFruit(e.target.value)}
        helpText="What do you prefer?"
        id="my-radio-help-text"
        label="Your favourite fruit"
        options={[
          {
            label: 'Apple',
            value: 'apple',
          },
          {
            label: 'Banana',
            value: 'banana',
          },
          {
            label: 'Grapefruit',
            value: 'grapefruit',
          },
        ]}
        value={fruit}
      />
    );
  }}
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <Radio
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-radio-success"
          label="Your favourite fruit"
          options={options}
          required
          validationState="valid"
          validationText="Great, they're in stock!"
          value={fruit}
        />
        <Radio
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-radio-warning"
          label="Your favourite fruit"
          options={options}
          required
          validationState="warning"
          validationText="Oh, really?"
          value={fruit}
        />
        <Radio
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-radio-error"
          label="Your favourite fruit"
          options={options}
          required
          validationState="invalid"
          validationText="You must select one kind of fruit."
          value={fruit}
        />
      </>
    );
  }}
</Playground>

### Disabled State

It's possible to disable just some options or the whole set.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        disabled: true,
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <Radio
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-radio-disabled-option"
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <Radio
          changeHandler={(e) => setFruit(e.target.value)}
          disabled
          id="my-radio-disabled"
          label="Your favourite fruit"
          options={options}
          value="apple"
        />
      </>
    );
  }}
</Playground>

## API

In addition to the options below, you can add any custom attributes that do not
interfere with the API, and they will be passed to the `input` HTML element.
This is useful mainly to improve component's accessibility.

<Props table of={Radio} />
