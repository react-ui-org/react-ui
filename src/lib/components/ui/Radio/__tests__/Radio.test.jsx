import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { Radio } from '../Radio';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<Radio
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly only one option', () => {
    const tree = shallow(<Radio
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly only one option and hidden label', () => {
    const tree = shallow(<Radio
      id="test"
      isLabelVisible={false}
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<Radio
      disabled
      helpText="some help"
      id="test"
      inFormLayout
      label="label"
      layout="horizontal"
      options={[
        {
          disabled: true,
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      required
      validationState="warning"
      validationText="some error"
      value="ch1"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler() on changing selected option', () => {
    const spy = sinon.spy();
    const component = mount(<Radio
      changeHandler={spy}
      id="id"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      value="ch2"
    />);

    component
      .find('input')
      .first()
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'ch1' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
