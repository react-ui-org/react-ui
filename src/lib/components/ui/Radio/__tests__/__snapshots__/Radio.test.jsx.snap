// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="root  rootLayoutVertical   "
>
  <div
    className="label "
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <ul
      className="list"
    >
      <li
        key="ch1"
      >
        <label
          className="option"
          htmlFor="test__item__ch1"
          id="test__item__ch1__label"
        >
          <input
            checked={false}
            className="input"
            id="test__item__ch1"
            name="test"
            onChange={null}
            type="radio"
            value="ch1"
          />
          <span
            className="optionLabel"
            id="test__item__ch1__labelText"
          >
            choice 1
          </span>
        </label>
      </li>
      <li
        key="ch2"
      >
        <label
          className="option"
          htmlFor="test__item__ch2"
          id="test__item__ch2__label"
        >
          <input
            checked={false}
            className="input"
            id="test__item__ch2"
            name="test"
            onChange={null}
            type="radio"
            value="ch2"
          />
          <span
            className="optionLabel"
            id="test__item__ch2__labelText"
          >
            choice 2
          </span>
        </label>
      </li>
    </ul>
  </div>
</div>
`;

exports[`rendering renders correctly only one option 1`] = `
<div
  className="root  rootLayoutVertical   "
>
  <div
    className="label "
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <ul
      className="list"
    >
      <li
        key="ch1"
      >
        <label
          className="option"
          htmlFor="test__item__ch1"
          id="test__item__ch1__label"
        >
          <input
            checked={false}
            className="input"
            id="test__item__ch1"
            name="test"
            onChange={null}
            type="radio"
            value="ch1"
          />
          <span
            className="optionLabel"
            id="test__item__ch1__labelText"
          >
            choice 1
          </span>
        </label>
      </li>
    </ul>
  </div>
</div>
`;

exports[`rendering renders correctly only one option and hidden label 1`] = `
<div
  className="root  rootLayoutVertical   "
>
  <div
    className="label isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <ul
      className="list"
    >
      <li
        key="ch1"
      >
        <label
          className="option"
          htmlFor="test__item__ch1"
          id="test__item__ch1__label"
        >
          <input
            checked={false}
            className="input"
            id="test__item__ch1"
            name="test"
            onChange={null}
            type="radio"
            value="ch1"
          />
          <span
            className="optionLabel"
            id="test__item__ch1__labelText"
          >
            choice 1
          </span>
        </label>
      </li>
    </ul>
  </div>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root isRootInFormLayout rootLayoutHorizontal isRootDisabled isRootRequired isRootStateWarning"
>
  <div
    className="label "
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <ul
      className="list"
    >
      <li
        key="ch1"
      >
        <label
          className="option"
          htmlFor="test__item__ch1"
          id="test__item__ch1__label"
        >
          <input
            checked={true}
            className="input"
            disabled={true}
            id="test__item__ch1"
            name="test"
            onChange={null}
            type="radio"
            value="ch1"
          />
          <span
            className="optionLabel"
            id="test__item__ch1__labelText"
          >
            choice 1
          </span>
        </label>
      </li>
      <li
        key="ch2"
      >
        <label
          className="option"
          htmlFor="test__item__ch2"
          id="test__item__ch2__label"
        >
          <input
            checked={false}
            className="input"
            disabled={true}
            id="test__item__ch2"
            name="test"
            onChange={null}
            type="radio"
            value="ch2"
          />
          <span
            className="optionLabel"
            id="test__item__ch2__labelText"
          >
            choice 2
          </span>
        </label>
      </li>
    </ul>
    <div
      className="helpText"
      id="test__helpText"
    >
      some help
    </div>
    <div
      className="validationText"
      id="test__validationText"
    >
      some error
    </div>
  </div>
</div>
`;
